### ISSUE TITLE
Allow to control order of extensions

### ISSUE THREAD

**Issue by Niklan** [thephpleague/commonmark#339]
It is possible to add extensions that will be prioritized over `CommonMarkCoreExtension`?

Maybe I didn't find the way to do this, but I need to add my parsers before default ones.

The case is, to define custom markdown syntax using chars from default extensions, but having different syntax.

## Case №1

We have `BlockQuoteParser` which triggers on first found `>` char without indent.

I want to implement syntax like from [MDSN](https://docs.microsoft.com/en-us/contribute/how-to-write-use-markdown#note-blocks) ([they use](https://github.com/MicrosoftDocs/Contribute/blob/master/Contribute/how-to-write-use-markdown.md#how-to-use-markdown-for-writing-docs) CommonMark compliant syntax):

```markdown
> [!NOTE]
> This is a NOTE

> [!WARNING]
> This is a WARNING
```

If I can put my parser before `BlockQuoteParser` somehow, I can check, if first line is continues with `[` or not, and return `FALSE`, to let default `BlockQuoteParser` do the job, or if I found it, handle rest of the work.

But I can't, because I `BlockQuoteParser` always returns true, and I can't place my block parser before it, so, my parser never will trigger.

## Case №2

Many of markdown documentations use [this YAML syntax](https://github.com/MicrosoftDocs/Contribute/edit/master/Contribute/how-to-write-use-markdown.md) ([another case](https://jekyllrb.com/docs/front-matter/)) to add meta-data to it.

```markdown
---
title: How to use Markdown for writing Docs
description: This article provides the basics and reference information for the Markdown language used for writing docs.microsoft.com articles.
ms.date: 01/29/2019
---
```

Also, this is not possible to implement, because `ThematicBreakParser` always returns `TRUE` on 3 or more `-`. And again, I can't insert my parser before it, so, it's never ben called. The only way here, is to add some text after it, like `---meta`, but this break parsing on GitHub, GitLab and other sites which supports this construction in Markdown.

## Possible solution

While I create experimental site, heavy using Markdown for content, I found, that I can't do many things, because default parsers mostly always returns `TRUE`. This is also the case for inline parsers.

Maybe, there is the reason to add `weight` property to parsers and set it to 0 by default to all custom parsers, -100 for default? An then, when `getBlockParsers()` (f.e.) called, they will be sorted by weight. Or add possibility to sort this list manually, by add method `setBlockParsers()`, in this way, we can get it, reorder as needed and then set it again.

### More info

I use [Drupal 8 module](https://www.drupal.org/project/markdown) listed in README.md.

They call [Environment::createCommonMarkEnvironment()](https://cgit.drupalcode.org/markdown/tree/src/Plugin/Markdown/LeagueCommonMark.php#n76), which instantly register default extensions:

```php
$environment->addExtension(new CommonMarkCoreExtension());
```

In my case, I can't even create fake $environment and add my extensions earlier. I can add patch to allow this in Drupal module, but think, this is must be done on library level.

_Sry for my English, hopes you understand me :sweat_smile:_

**Comment by colinodell** [comment]
> It is possible to add extensions that will be prioritized over CommonMarkCoreExtension?
>
> Maybe I didn't find the way to do this, but I need to add my parsers before default ones.

You are correct that the current implementation relies on the order extensions are added.  You would need to create an empty `Environment` first, add your custom extension, and then add the default one.

I do like your idea of using a `weight` or `priority` similar to Symfony's event listener:

![image](https://user-images.githubusercontent.com/202034/53699101-bc0dba00-3db2-11e9-8890-0c1bd36aeb2b.png)

For the `Environment` class, we can add a new parameter `$priority = 0` to each of the `add___()` methods:

```diff
     /**
      * @param BlockParserInterface $parser
+     * @param int                  $priority
      *
      * @return $this
      */
-    public function addBlockParser(BlockParserInterface $parser)
+    public function addBlockParser(BlockParserInterface $parser, $priority = 0)
     {
         $this->assertUninitialized('Failed to add block parser.');
-        $this->getMiscExtension()->addBlockParser($parser);
+        $this->getMiscExtension()->addBlockParser($parser, $priority);
         return $this;
     }
```

This would be a minor BC-break as it only impacts people who extend this class, which is probably rare.

The bigger question is how we should change the `ExtensionInterface` (and its implementations) to indicate the priority of the parsers, renderers, etc.  Because this would require a larger BC-break, I'm okay if we want to take this opportunity to completely change the interface to something else.  For example, instead of having the individual `get___()` methods, maybe we implement a single `register(Environment $env)` method and have the extensions add their individual components themselves?  I don't know if I like this idea, or how well it might work - it's just an example of the kinds of change we should evaluate at this time.

If you (or anyone else watching this repo) have a more specific suggestion about how you think the `Environment` and `ExtensionInterface` should be changed then let me know - I'd love to hear it!

> Sry for my English, hopes you understand me

Your English is very good - no need to apologize :smile: 

Thank you so much for the suggestion and the detailed write-up!

**Comment by colinodell** [comment]
We might also want to revisit [#35 Support for multiple renderers](https://github.com/thephpleague/commonmark/issues/35) at this time

**Comment by Niklan** [comment]
Do we actually need to pass `priority` via method?

Maybe it's better to add to `AbstractBlockParser` property and getters\setters.

```diff
abstract class AbstractBlockParser implements BlockParserInterface
{

+   /**
+    * The parser priority over others.
+    *
+    * @var int
+    */
+   protected $priority = 0;

    /**
     * @return string
     */
    public function getName()
    {
        $reflection = new \ReflectionClass($this);
         return $reflection->getShortName();
    }

+   /**
+    * {@inheritdoc}
+    */
+   public function getPriority() {
+     return $this->priority;
+   }

+   /**
+    * {@inheritdoc}
+    */
+   public function setPriority($priority) {
+     $this->priority = $priority;

+     return $this;
+   }

}
```

This way, all defined parser will have priority 0 by default, and doesn't brake anything at all.

I didn't fully understand where the parsers finally sorted in object, but thinks there is place to sort all of them here.

I also see problem with priority via argument. There is possibility to add parsers via `Extension` implementation and Environment `$environment->addExtension(new MyExtensions());`, how in this way ppl will pass priority? For me, it's much safer handle and store priorty inside individual parser, not in method.

Also, I have another idea, which is similar to described above. This is same technique, but using static method. Like so

```php
/**
 * {@inheritdoc}
 */
public static function getPriority(Environment $environment) {
  return 0;
}
```

And call it when final parser list is prepared. We can pass some additional information such as `Environment` and other stuff, to make decision, how much priority this parser must get.

If I get it right, we just need to add sorting logic inside `Environment::getBlockParsers()` something like [this](https://github.com/linnovate/doil/blob/master/core/lib/Drupal/Component/Datetime/DateTimePlus.php).

---

Both solution can work out of the box, without breaking any of implementations, because we can set default value in `AbstractBlockParser` which as I understand, by every parser instance.

**Comment by colinodell** [comment]
> Do we actually need to pass priority via method?
> 
> Maybe it's better to add to AbstractBlockParser property and getters\setters.

I want to make it easy for users to adjust the priority should they need to, while still allowing this library and other extensions to specify their recommended priorities.

Technically I'd like the priorities set per-parser, not per-extension. Extensions are just a way to simplify the registration of related parsers, processors, and renderers with sane defaults.  I feel like you already understand that, I just want to drive that point home :)

> This is just an example that this is possible even without any BC changes

Thank you for considering BC changes!  They are definitely something I want to minimize as much as possible, although I'm very open to breaking BC provided that:

 - The changes greatly improve this library and are something we'd want in a v1.0 release
 - The impact is minimized where possible
 - The upgrade process is documented in UPGRADE.md and fairly straight-forward to implement

In this case, I think adding a new "optional" parameter `$priority = 0` to each of the `add___()` methods meets those goals.

> This sorting results must be cached(?) or I use wrong place to impelement sorting logic

Yeah, the `Environment` uses this concept of "initialization". Basically, you must `add` all your extensions, parsers, etc. first.  As soon as you attempt to "use" the `Environment` (by calling any non-`add___()` methods), we perform the "initialization" step where we finalize the `Environment`.  This step would be the perfect time to perform the sort and cache the results.

Speaking of sorting, I noticed you were using PHP 5.6-compatible code in `uasort`.  I'm okay if we want to use this opportunity to drop PHP 5.6 support so we can leverage the spaceship operator :)

**Comment by colinodell** [comment]
I've been hacking away at this today and I think I've got an implementation that I like.  I'll push up a PR so we can review it together and gather some feedback :)

**Comment by colinodell** [comment]
@Niklan Please review #341 when you get a chance - I'd love to hear your feedback!

**Comment by Niklan** [comment]
Cool! I'll definitely look at PR a bit later. And test my cases :+1:

**Comment by colinodell** [comment]
Implemented via #341
